### ISSUE TITLE
phase correct failure (nan related)

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#149]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/149) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Here is a minimum script that causes the failure

```
import mass
import numpy as np
import pylab as plt
import os
from matplotlib.backends.backend_pdf import PdfPages
import argparse
import h5py
import sys
import devel
from devel import Side, Sides, RepeatedLinePlotter, WorstSpectra
import pickle

data = mass.TESGroup(['/Volumes/HHD3/horton/20171222_B/20171222_B_chan1.ljh'],['/Volumes/HHD3/horton/20171222_A/20171222_A_chan1.ljh'])
fname = data.hdf5_file.filename[:]
data.hdf5_file.close()
if os.path.isfile(fname): os.remove(fname)
data = mass.TESGroup(['/Volumes/HHD3/horton/20171222_B/20171222_B_chan1.ljh'],['/Volumes/HHD3/horton/20171222_A/20171222_A_chan1.ljh'])
data.updater = mass.core.utilities.NullUpdater
data.summarize_data(forceNew=True)
data.auto_cuts()
data.avg_pulses_auto_masks(forceNew=True)
data.compute_filters(forceNew=True)
data.filter_data(forceNew=True)
data.drift_correct(forceNew=True)
ds = data.first_good_dataset
ds.phase_correct()
```

The output looks like
```
AssertionError: 
> /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.6.2-py2.7-macosx-10.12-x86_64.egg/mass/core/channel.py(2340)_phase_corrected_filtvals()
   2338 
   2339     ph = np.hstack([0] + [c(0) for c in corrections])
-> 2340     assert (ph[1:] > ph[:-1]).all()  # corrections should be sorted by PH
   2341     corr = np.zeros((NC+1, NP), dtype=float)
   2342     for i, c in enumerate(corrections):

ipdb> 
```

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

It won't let me attach a file larger than 100 MB :(. So you can get the file from `pcuser@687gombp:/Volumes/HHD3/horton/20171222_B/20171222_B_chan1.ljh` for now.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

I can replicate your PDB `AssertionError`, and I think it gives me some clues about what to do. 

On the other hand, I don't see anything about a NaN. So for example,

```
ipdb> np.isnan(ph)
array([False, False, False, False, False, False, False, False], dtype=bool)
ipdb> np.any(np.isnan(ph))
False
```

Is it possible that "nan related" is not quite right? The problem instead seems to be that it tried to make corrections at PH of 0 and also at 0, which fails.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Okay, this is crazy, but there are TWO problems, and I think you have to fix both:

1. Your data contains a large peak at approximately 200 in `p_filt_value_dc` units. That is baloney. You need to cut it with `for ds in data: ds.cuts.cut("pulse_average", ds.p_filt_value_dc[:]<1000)` or something like that. If you don't, the `phase_correct()` method will try to correct that peak for arrival-time systematics. That's the source of the problem that you outlined above.

1. But also there's an error that I see not infrequently.  Do you recognize this?

```
In [21]: ds.phase_correct(forceNew=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-21-52d5b3f555d9> in <module>()
----> 1 ds.phase_correct(forceNew=True)

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.6.2-py2.7-macosx-10.12-x86_64.egg/mass/core/channel.pyc in phase_correct(self, forceNew, category, ph_peaks, method2017, kernel_width)
   1599                                                   .012*np.mean(ph_peaks),
   1600                                                   method2017=method2017,
-> 1601                                                   kernel_width=kernel_width)
   1602             corrections.append(c)
   1603             median_phase.append(mphase)

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.6.2-py2.7-macosx-10.12-x86_64.egg/mass/core/channel.pyc in _phasecorr_find_alignment(phase_indicator, pulse_heights, peak, delta_ph, method2017, nf, kernel_width)
   2351     for i, P in enumerate(Pctrs):
   2352         use = (Pbin == i)
-> 2353         c, b = np.histogram(pulse_heights[use], NBINS, phrange)
   2354         hists[i] = c
   2355     bctr = 0.5*(b[1]-b[0])+b[:-1]

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/numpy/lib/function_base.pyc in histogram(a, bins, range, normed, weights, density)
    778             else:
    779                 n += np.bincount(indices, weights=tmp_w,
--> 780                                  minlength=bins).astype(ntype)
    781 
    782         # Rename the bin edges for return.

ValueError: The first argument of bincount must be non-negative
> /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/numpy/lib/function_base.py(780)histogram()
    778             else:
    779                 n += np.bincount(indices, weights=tmp_w,
--> 780                                  minlength=bins).astype(ntype)
    781 
    782         # Rename the bin edges for return.
```

This apparently happens because `p_filt_value_dc` is of type `np.float32`, but the function `np.histogram(...)` operates in a funny way. It figures out the bin edges and discards all values that lie outside the range. Then it converts all surviving data into a **`float`** bin number by subtracting the lowest value and dividing by the per-bin range. But `float==np.float64`, so a number that is represented as equal to the lowest bin edge in float32 might turn out to lie below that edge after conversion to float 64. Got all that? If not, try the following trick:

```
In [187]: a=np.array([20752.0546875,  21194.875], dtype=np.float32); mn=20752.0554691

In [188]: a-mn
Out[188]: array([   0.       ,  442.8203125], dtype=float32)

In [189]: a.astype(float)-mn
Out[189]: array([ -7.81600000e-04,   4.42819531e+02])
```

See how the first value is "equal to" `mn` in 32-bit land, but it's less than `mn` in 64-bit land? I can't decide what to do here. It seems like a bug in `numpy`, honestly. I'll search for bug reports along these lines.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

I submitted [numpy issue 10319](https://github.com/numpy/numpy/issues/10319) about this matter. But that will take time to fix, if the numpy community even agrees to fix it at all. I'll think about a workaround purely within mass today.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Thanks for the info. The second thing is very weird, and doesn't happen in julia. I'm not sure what is going on, but I think that backs up your assertion that it is a numpy bug.

```julia 
 julia> a = Float32[20752.0546875,  21194.875]
2-element Array{Float32,1}:
 20752.1
 21194.9

julia> mn = 20752.0554691
20752.0554691

julia> a-mn
2-element Array{Float64,1}:
  -0.0007816
 442.82 
```

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I'm guessing if we converted to Float64 ourselves before histogramming it would be fine. Is there a reason this happened in this data specifically?

Also I wrote nan related because in my case `ph[0]` was nan, maybe we have different versions of numpy? (I'm on 1.13.3, you are on 1.13.1 according to your issue). I don't think that difference is worth understanding though.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

You misunderstand. Your julia code does exactly what python does. The problem is there in that negative first value in `a-mn`. The numpy histogram function converts that to integer bin number -1, as it ought to, but the bin indices  are then passed to `np.bincount`, which cannot handle negative bin indices and consequently raises `ValueError`.

The histogram function assumes that all values that would have given `index < 0` or `index >= numbins` have already been removed at this point. Unfortunately, the selection that removes the out-of-limits values is only *approximately* correct when the input values are float32s.

We can convert to float64 before histogramming, but that's a little ugly and somewhat expensive in some cases. One problem is that this surely happens in other spots besides `ds.phase_correct`. Another is that we might have huge arrays of float32 in our HDF5 file, and there's no need to convert the entire array if we're only going to include a small fraction in the histogram limits.

Proposal: we could add something like the following to MASS:

```python
def safe_histogram(a, bins, bin_limits):
    """Workaround for numpy bug in histogramming float32 data. See https://github.com/numpy/numpy/issues/10319"""
    if a.dtype == np.float32:
        keep = a >= bin_limits[0]
        keep &= a < bin_limits[1]
        return np.histogram(a[keep].astype(np.float), bins, bin_limits)
    return np.histogram(a, bins, bin_limits)
```

We would call this instead of `np.histogram` in any places where we found this error arises.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

In Julia you ALWAYS get a negative value, you never get 0. Thats why I said it doesn't happen. My example was analogous to your line 188 which results in `array([   0.       ,  442.8203125], dtype=float32)` in python, but `Float64[-0.0007816 442.82]` in julia.

I think your proposal is pretty good, it's simple and universal, and probably easy to do with search and replace for `np.nistogram(` -> `safe_histogram(`.  Except that I often use the form `np.histogram(data,np.arange(100))` where you don't pass a number of bins and limits separately.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Actually why don't we write a new version of `numpy.histogram` as in your pull request, then monkey patch it in?

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

We could, but (a) I don't have a pull request, just an issue, and (b) they marked it as a duplicate. Also, they appear to have completely re-factored `numpy.histogram` in recent versions. That means that patching in the latest is more than just replacing one function.

As to monkey-patching, I'm not sure how we (MASS) would know *when* to do it. Should we test on numpy's version number? Or perhaps we run my test case and patch if it raises `ValueError`? It's ugly, but it might be better than trying to wrap all our own calls of `np.histogram`.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

After we discussed over lunch, the plan is:

* Make a test that fails.
* Run a similar test at mass-import time. If it fails, monkey-patch a replacement that wraps `np.histogram`.

This is a little scary, but given that the problem is so widespread and even built into matplotlib, too, I think we need to monkey-patch.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

A work-around for a bug in numpy.histogram in numpy 1.13.

Monkey-patches np.histogram to use our safer version, avoiding a bug
when the input data array is a float32 or float16. If the imported
numpy doesn’t show the bug, then don’t monkey patch. Fixes #149.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I'm currently testing your pull request, but I just was confused by `for ds in data: ds.cuts.cut("pulse_average", ds.p_filt_value_dc[:]<1000)` it using pulse average on the left, but filt_value_dc on the right so I kept cutting all my pulses by running this before drift correction. Not a complaint, I thought you might find it amusing.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Merged in fowler (pull request #49)

Fix #149 by working around a bug in numpy.

Approved-by: Galen O'Neil <galen.oneil@nist.gov>

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.7 (automated comment)
